test_vm/src/lib.rs

```diff
@@ -606,6 +812,115 @@ pub fn actor(code: Cid, head: Cid, seq: u64, bal: TokenAmount) -> Actor {
+    pub code: Option<ExitCode>,
+    pub ret: Option<RawBytes>,
+    pub subinvocations: Vec<InvocationTrace>,
+}
+
+pub struct ExpectInvocation {
+    pub to: Address,       // required
+    pub method: MethodNum, // required
+    pub code: Option<ExitCode>,
+    pub from: Option<Address>,
+    pub params: Option<RawBytes>,
+    pub ret: Option<RawBytes>,
+    pub subinvocs: Option<Vec<ExpectInvocation>>,
+}
+
+impl ExpectInvocation {
+    // testing method that panics on no match
+    pub fn matches(&self, invoc: &InvocationTrace) {
+        let id = format!("[{}:{}]", invoc.msg.to, invoc.msg.method);
+        self.quick_match(invoc, "".to_string());
+        if let Some(c) = self.code {
+            assert_ne!(
+                None,
+                invoc.code,
+                "{} unexpected code: expected:{}was:{}",
+                id,
+                c,
+                ExitCode::OK
+            );
+            assert_eq!(
+                c,
+                invoc.code.unwrap(),
+                "{} unexpected code expected:{}was:{}",
+                id,
+                c,
+                invoc.code.unwrap()
+            );
+        }
+        if let Some(f) = self.from {
+            assert_eq!(
+                f, invoc.msg.from,
+                "{} unexpected from addr: expected:{}was:{} ",
+                id, f, invoc.msg.from
+            );
+        }
+        if let Some(p) = &self.params {
+            assert_eq!(
+                p, &invoc.msg.params,
+                "{} unexpected params: expected:{:x?}was:{:x?}",
+                id, p, invoc.msg.params
+            );
+        }
+        if let Some(r) = &self.ret {
+            assert_ne!(None, invoc.ret, "{} unexpected ret: expected:{:x?}was:None", id, r);
+            let ret = &invoc.ret.clone().unwrap();
+            assert_eq!(r, ret, "{} unexpected ret: expected:{:x?}was:{:x?}", id, r, ret);
+        }
+        if let Some(expect_subinvocs) = &self.subinvocs {
+            let subinvocs = &invoc.subinvocations;
+
+            let panic_str = format!(
+                "unexpected subinvocs:\n expected: \n[\n{}]\n was:\n[\n{}]\n",
+                self.fmt_expect_invocs(expect_subinvocs),
+                self.fmt_invocs(subinvocs)
+            );
+            assert!(subinvocs.len() == expect_subinvocs.len(), "{}", panic_str);
+
+            for (i, invoc) in subinvocs.iter().enumerate() {
+                let expect_invoc = expect_subinvocs.get(i).unwrap();
+                // only try to match if required fields match
+                expect_invoc.quick_match(invoc, panic_str.clone());
+                expect_invoc.matches(invoc);
+            }
+        }
+    }
+
+    pub fn fmt_invocs(&self, invocs: &[InvocationTrace]) -> String {
+        let mut out = "".to_string();
+        for (i, invoc) in invocs.iter().enumerate() {
+            out = format!("{}{}: [{}:{}],\n", out, i, invoc.msg.to, invoc.msg.method);
+        }
+        out
```

Comment:
```suggestion
    invocs.iter().enumerate().map(|(i, invoc) {
        format!("{}{}: [{}:{}],\n", out, i, invoc.msg.to, invoc.msg.method)
    }).collect()
```

Reply:
(and below)

Reply:
Is var `out` and first `{}` needed at all after this?

Reply:
collect is still a mystery to me, is it acting like reduce with accumulation function string concat?

Reply:
> collect is still a mystery to me, is it acting like reduce with accumulation function string concat?

Yes. `collect` is basically: take this iterator, and smoosh it into this collection-like thing. Usually, this is for collecting an iterator into something like a vec. But you can also collect strings into a single string (concatenating them).